### PR TITLE
Add hard/soft reset methods

### PR DESCRIPTION
**Pull request by pkendall64**
Add `hard_reset` and `soft_reset` methods to the public API for ESPLoader.

### CODE REVIEW

File: ESPLoader.js [hunk 1]
```diff
@@ -1586,6 +1587,29 @@ class ESPLoader {
...
         this.log("Detected flash size: " + this.DETECTED_FLASH_SIZES[flid_lowbyte]);
     }
+
+    hard_reset = async() => {
+        this.transport.setRTS(true);  // EN->LOW
+        await this._sleep(100);
+        this.transport.setRTS(false);
+    }
+
+    soft_reset = async() => {
```
Comment (balloob) [inline ESPLoader.js:1597]: When is a soft reset used versus a hard reset?

File: ESPLoader.js [hunk 2]
```diff
@@ -1586,6 +1587,29 @@ class ESPLoader {
...
+        if (!this.IS_STUB) {
+            // 'run user code' is as close to a soft reset as we can do
```
Comment (balloob) [inline ESPLoader.js:1599]: If we offer an API it should do the thing, not be "as close to". So in that case this method should raise unless it's ESP8266 + Stub.

File: ESPLoader.js (hunk 1 above)
Comment (pkendall64) [reply ESPLoader.js:1597]: A soft reset is used when the circuit doesn't utilise the RTS/DTR lines. As in an ExpressLRS receiver.

File: ESPLoader.js (hunk 2 above)
Comment (pkendall64) [reply ESPLoader.js:1599]: The comment comes directly from `esptool.py`. The reason is that in the ROM bootloader there is no way to perform a reset and so just running the user code is as close to a reset as you can get given the limitations.
Perhaps the comment should be
```
// The ROM bootloader has to reset function so we just run the user code
```

Comment (balloob) [reply ESPLoader.js:1599]: Didn't realize this was ported from ESPTool 👍

File: ESPLoader.js [hunk 3]
```diff
@@ -1586,6 +1587,29 @@ class ESPLoader {
...
+            if (this.chip.CHIP_NAME != "ESP8266") {
+                throw("Soft resetting is currently only supported on ESP8266");
+            } else {
+                // running user code from stub loader requires some hacks
+                // in the stub loader
+                this.command({op: this.ESP_RUN_USER_CODE, wait_response: false});
+            }
+        }
+
+    }
```
Comment (balloob) [inline ESPLoader.js:1612]: Merge else + if.
```suggestion
        } else if (this.chip.CHIP_NAME != "ESP8266") {
            throw("Soft resetting is currently only supported on ESP8266");
        } else {
            // running user code from stub loader requires some hacks
            // in the stub loader
            this.command({op: this.ESP_RUN_USER_CODE, wait_response: false});
        }
    }
```

File: ESPLoader.js [hunk 4]
```diff
@@ -1586,6 +1587,29 @@ class ESPLoader {
...
+
+    soft_reset = async() => {
+        if (!this.IS_STUB) {
+            // 'run user code' is as close to a soft reset as we can do
+            this.flash_begin(0, 0);
+            this.flash_finish(false);
+        } else {
+            if (this.chip.CHIP_NAME != "ESP8266") {
+                throw("Soft resetting is currently only supported on ESP8266");
```
Comment (balloob) [inline ESPLoader.js:1604]: This should be `throw ESPError` when #26 is merged.

File: ESPLoader.js [hunk 5]
```diff
@@ -1586,6 +1587,26 @@ class ESPLoader {
...
+        this.transport.setRTS(false);
+    }
+
+    soft_reset = async() => {
+        if (!this.IS_STUB) {
+            // 'run user code' is as close to a soft reset as we can do
+            this.flash_begin(0, 0);
+            this.flash_finish(false);
+        } else if (this.chip.CHIP_NAME != "ESP8266") {
+            throw("Soft resetting is currently only supported on ESP8266");
```
Comment (balloob) [inline ESPLoader.js:1603]: Error reporting #26 has been merged. Shall we have this raise?
```suggestion
            throw new ESPError("Soft resetting is currently only supported on ESP8266");
```

**Comment by adwait-esp** [comment]
LGTM

**Pull request merged**